malicious-site-protection/malicious-site-protection-impl/src/main/kotlin/com/duckduckgo/malicioussiteprotection/impl/data/db/MaliciousSiteDao.kt

```diff
@@ -0,0 +1,123 @@
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insertRevision(revision: RevisionEntity)
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insertHashPrefixes(items: List<HashPrefixEntity>)
+
+    @Query("DELETE FROM hash_prefixes")
+    suspend fun deleteHashPrefixes()
+
+    @Query("DELETE FROM filters")
+    suspend fun deleteFilters()
+
+    @Query("DELETE FROM hash_prefixes WHERE type = :type")
+    suspend fun deleteHashPrefixes(type: String)
+
+    @Query("DELETE FROM hash_prefixes WHERE hashPrefix = :hashPrefix AND type = :type")
+    suspend fun deleteHashPrefix(
+        hashPrefix: String,
+        type: String,
+    )
+
+    @Query("DELETE FROM filters WHERE hash = :hash AND type = :type")
+    suspend fun deleteFilter(
+        hash: String,
+        type: String,
+    )
+
+    @Query("DELETE FROM filters WHERE type = :type")
+    suspend fun deleteFilters(type: String)
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insertFilters(items: Set<FilterEntity>)
+
+    @Query("SELECT * FROM revisions")
+    suspend fun getLatestRevision(): List<RevisionEntity>?
+
+    @Query("SELECT * FROM revisions WHERE feed = :feed AND type = :type")
+    suspend fun getLatestRevision(feed: String, type: String): RevisionEntity?
+
+    @Query("DELETE FROM revisions")
+    suspend fun deleteRevisions()
+
+    @Query("SELECT * FROM hash_prefixes WHERE hashPrefix = :hashPrefix")
+    suspend fun getHashPrefix(hashPrefix: String): HashPrefixEntity?
+
+    @Query("SELECT * FROM filters WHERE hash = :hash")
+    suspend fun getFilter(hash: String): List<FilterEntity>?
+
+    @Transaction
+    suspend fun updateHashPrefixes(
+        hashPrefixes: HashPrefixesWithRevision?,
+    ) {
+        hashPrefixes ?: return
+
+        val currentLocalRevision = getLatestRevision(feed = hashPrefixes.feed.name, type = hashPrefixes.type.name)?.revision ?: 0
+        val newRevision = hashPrefixes.takeIf { it.revision > currentLocalRevision }?.apply {
+            Timber.d("Cris. Updating hash prefixes")
+            if (hashPrefixes.replace) {
+                deleteHashPrefixes(hashPrefixes.feed.name)
+            } else {
+                hashPrefixes.delete.forEach {
+                    deleteHashPrefix(it, hashPrefixes.feed.name)
+                }
+            }
+            insertHashPrefixes(hashPrefixes.insert.map { HashPrefixEntity(hashPrefix = it, type = hashPrefixes.feed.name) })
+        }?.revision ?: currentLocalRevision
+        if (currentLocalRevision != newRevision) {
+            insertRevision(RevisionEntity(feed = hashPrefixes.feed.name, type = hashPrefixes.type.name, revision = newRevision))
+        }
+    }
+
+    @Transaction
+    suspend fun updateFilters(
+        filterSet: FilterSetWithRevision?,
+    ) {
+        filterSet ?: return
+
+        val currentLocalRevision = getLatestRevision(feed = filterSet.feed.name, type = filterSet.type.name)?.revision ?: 0
```

Comment:
NIT: there's repeated code on this logic and the previous method. We can extract a generic method to simplify it, and if we ever touch the logic (for a bug or improvement), we will do in a single place.